### PR TITLE
#627 - Retrieval of existing annotations in PDF editor frontend

### LINKED ISSUE #627: Retrieval of existing annotations in PDF editor frontend
**Is your feature request related to a problem? Please describe.**
Already existing annotations must be rendered when opening the PDF editor.

**Describe the solution you'd like**
PDF editor sends an ajax request to the backend for receiving existing annotations.
Annotations are fetched in the backend and converted to a suitable representation for PDFAnno .
Then the annotations are sent to the frontend and rendered in PDFAnno.

### PR DESCRIPTION
**Pull request by wpmp**
**What's in the PR**
* Adds support for retrieval existing annotations in the PDF editor frontend, closes #627

**How to test manually**
* add a PDF document
* add annotations the PDF document in brat editor
* open PDF document in PDF editor

**Automatic testing**
* [X] PR includes unit tests

**Documentation**
* [ ] PR updates documentation

### CODE REVIEW

**Comment by wpmp** [comment]
Jenkins, can you test this please?

File: inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js [hunk 1]
```diff
@@ -159,6 +160,14 @@ async function displayViewer () {
             window.annoPage.scrollToAnnotation(moveTo)
           }, 500)
         }
+*/
+        parent.Wicket.Ajax.ajax({
+          "m" : "GET",
+          "u" : annoURL,
+          "sh" : [],
+          "fh": []
```
Comment (reckart) [inline inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: I guess it would be a good idea to display an error message if there was a failure - at least to log one to the JS console.

**Comment by reckart** [comment]
What do you mean by this:

> Append a pdfanno annotation to the return string of method renderAnnoFile in PdfAnnotationEditor class

**Comment by reckart** [comment]
IMHO #627 is not closed unless annotations are actually retrieved from the CAS and rendered - or do we have another issue for that?

**Comment by wpmp** [comment]
There is another issue for getting annotations in backend ( #630 ).
I expected this to be a bit more work for the frontend.
Should I maybe close this PR and commit these changes together with backend stuff in another PR and branch?
I did frontend first so see what format I would neeed and then thought I could commit it directly.

**Comment by wpmp** [comment]
Alright, so I continue on this branch and PR and add backend functionality here.
Should I somehow mention #630 in #627?

**Comment by reckart** [comment]
Best close #630 with a comment that it is handled together with #627 and update the description of #627 accordingly.

**Comment by wpmp** [comment]
@reckart I added rendering of all span annotations in pdf editor at once.
This includes rendering colors in the simple way of using color string as label and having an identity color map for now.
I think this should be fine, if I find any conflicts I will create new issues.

What is not included is the following:
* handling of ligatures
* rendering of only loaded pages
* rendering of relations
* handling of discontinued annotations

I will create new issues for those.

**Comment by wpmp** [comment]
I will add some tests. :)

**Comment by wpmp** [comment]
Jenkins, can you test this please?

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/PdfAnnotationEditor.java [hunk 2]
```diff
@@ -74,4 +83,24 @@ private void handleError(String aMessage, Throwable aCause)
...
+     * @param pdftxt Output string of PDFExtract
+     * @return Annotation file
+     */
+    public AnnoFile renderAnnoFile(String pdftxt) throws IOException
+    {
+        JCas jCas = getJCasProvider().get();
+        if (getModelObject().getProject() != null)
+        {
+            PdfExtractFile pdfExtractFile = new PdfExtractFile(pdftxt);
+            VDocument vdoc = render(jCas, 0, Integer.MAX_VALUE);
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/PdfAnnotationEditor.java]: Use `jCas.getDocumentText().length()` instead of `Integer.MAX_VALUE`.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/PdfAnnotationEditor.java [hunk 3]
```diff
@@ -74,4 +83,24 @@ private void handleError(String aMessage, Throwable aCause)
...
+            AnnoFile annoFile = PdfAnnoRenderer.render(getModelObject(),
+                vdoc, jCas.getDocumentText(), annotationService, pdfExtractFile);
+            return annoFile;
+        }
+
+        throw new IOException("");
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/PdfAnnotationEditor.java]: Provide a proper error message. Why is this an IOException anyway - there is no IO here? I'd probbaly simply do nothing if the project is null?

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoPanel.java [hunk 4]
```diff
@@ -101,6 +107,41 @@ public void onRequest()
             }
         });
 
+        add(annoProvider = new AbstractDefaultAjaxBehavior() {
+
+            private static final long serialVersionUID = 3L;
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoPanel.java]: Better use generated serial IDs.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoPanel.java [hunk 5]
```diff
@@ -101,6 +107,41 @@ public void onRequest()
...
+                        "'primary': true," +
+                        "'colorMap': " + annoFile.getColorMap().toString() + "," +
+                        "'annotations':[annoFile]}, true);" +
+                        "}, 10);";
+
+                    aTarget .appendJavaScript(script);
+                }
+                catch (IOException e)
+                {
+                    log.error("Unable to get PDF text for " + pdfFile.getName()
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoPanel.java]: Also add a use-visible error message by calling `org.apache.wicket.Component.error(Serializable)` and `target.addChildren(getPage(), IFeedback.class);`.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/AnnoFile.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.tudarmstadt.ukp.inception.pdfeditor.pdfanno.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class AnnoFile
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/AnnoFile.java]: Please add JavaDoc.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/AnnoFile.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    private List<Span> spans;
+
+    private List<Relation> relations;
+
+    private ColorMap colorMap;
+
+    public AnnoFile(String pdfannoVersion, String pdfextractVersion)
+    {
+        this.pdfannoVersion = pdfannoVersion;
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/AnnoFile.java]: Use `aPdfAnnoVersion` and `aPdfExtractVersion` and remove the `this.`.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/AnnoFile.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        }
+    }
+
+    public ColorMap getColorMap()
+    {
+        return colorMap;
+    }
+
+    @Override
+    public String toString()
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/AnnoFile.java]: It looks like the `toString()` method is used to serialize the file to a string. Usually `toString()` is used for debugging. Better use a different method name. If you want toString() to return the string-serialized version, then call that new method in toString().

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/Annotation.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.tudarmstadt.ukp.inception.pdfeditor.pdfanno.model;
+
+public interface Annotation {
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/Annotation.java]: What is the purpose?

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/ColorMap.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        return relation;
+    }
+
+    public void addRelation(String name, String color)
+    {
+        relation.put(name, color);
+    }
+
+    private String mapToString(Map<String, String> map)
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/ColorMap.java]: When dealing with JSON, better use Jackson to map between Java <-> JSON instead of hand-cobbling the JSON string together.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+    private String stringContent;
+
+    /**
+     * Contains position mapping for a character between PDFExtract string
+     * including and excluding Draw Operations.
+     */
+    private Map<Integer, Integer> stringPositionMap;
+
+    private Map<Integer, PdfExtractLine> extractLines;
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java]: "extractedLines"? Please add JavaDoc.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+    public PdfExtractFile(String pdftxt)
+    {
+        setPdftxt(pdftxt);
+    }
+
+    public void setPdftxt(String pdftxt)
+    {
+        this.stringPositionMap = new HashMap<>();
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java]: Please prefix arguments with `a` and do not use `this.`. Applies in general.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    public String getStringContent()
+    {
+        return stringContent;
+    }
+
+    /**
+     * Gets PdfExtractLines between the given range in the string-only content
+     * @param start
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java]: Either describe the parameters/return value or remove the stub `@param/@return` lines having no descriptions. Applies in general.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractLine.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2017
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractLine.java]: Year ;) Applies to all new files.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/Relation.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        return tail;
+    }
+
+    public String getLabel()
+    {
+        return label;
+    }
+
+    @Override
+    public String toString()
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/Relation.java]: See previous comment on serializing via toString().

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractLine.java (hunk 14 above)
Comment (wpmp) [reply inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractLine.java]: Is it 2018 or 2019 now? :)

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java (hunk 12 above)
Comment (wpmp) [reply inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java]: Would it be `anId` then or should I go with `a` for all variables?

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractLine.java (hunk 14 above)
Comment (reckart) [reply inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractLine.java]: Well, definitely not 2017 ;)

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java (hunk 12 above)
Comment (reckart) [reply inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java]: "a" for "argument" not for "a" (determiner) - so not "anId" but "aId"

**Comment by reckart** [comment]
> @reckart yes, as I said, render annotations are not included right now.
I opened an issue for it and thought I'd add it in another PR to get this one closed as it becomes quite large, what do you think?

I agree.

**Comment by reckart** [comment]
I get warnings in the console like this:

```
2019-01-03 23:36:22 WARN [SYSTEM] COSDocument - Warning: You did not close a PDF Document
```

Also, I do not see any annotations in the PDF view anymore. 

Tested with: 59aba38cdef284a92f7e5174a975ae4920145114

**Comment by wpmp** [comment]
For me annotations work just fine.
However I get the warning, too.
It seems to be related to PDFBox ( https://pdfbox.apache.org/1.8/faq.html#notclosed ).
Maybe the usage of PDFExtract is introducing this warning.
I will have a look.

**Comment by reckart** [comment]
@wpmp can you please check that the commit you test with is 55bb2eff8eca98cf67b8cca59e68db8cbb9347c0 and that you clear your browser cache before accessing the annotation screen? Do you still see the annotations?

**Comment by wpmp** [comment]
@reckart am I missing something?
55bb2ef is a commit from another PR, isn't it?
How can I test this PR together with PR for #776?
Do I need to merge the branches in a new branch which I delete afterwards?

**Comment by wpmp** [comment]
No problem, maybe it was just too late that day :D
About the warnings: I open a PR to PDFExtract where I close the PDDocuments after usage.
It seems warnings disappear with this fix.
So we need a new release for PDFExtract then.

**Comment by wpmp** [comment]
@reckart what do you mean? For me relations are drawn as lines in PDFAnno, they get highlighted if I hover over the dots or over the line.
Maybe the color is too bright?

**Comment by reckart** [comment]
In the brat view, there is a relation (or more precisely a chain link) between the span annotations on "Anonymous" and "Submission". When I move the mouse there, only the circles react, but no line is drawn. Between "Unboxed" and "models", there is no relation/chain link. I had tried also with a dependency link between the two POS spans on "Unboxed" and "models", but the result was the same: only the blue circles reacted but no line was drawn.

**Comment by wpmp** [comment]
That is really weird. I can't reproduce, for me it shows the arrow lines for chains and relations... :/

**Comment by reckart** [comment]
@wpmp For me, it works in Chrome (Version 71.0.3578.98 (Official Build) (64-bit)) but it does not work in Safari (Version 12.0.2 (13606.3.4.1.4)).

**Comment by wpmp** [comment]
I am using Chrome, same version and I have no access to Safari.
I think PDFAnno only supports Chrome.
Can you make relations in PDFAnno? (Demo: https://paperai.github.io/pdfanno/latest/)

**Comment by reckart** [comment]
I have opened an issue upstream: https://github.com/paperai/pdfanno/issues/196
